manual/src/refman/extensions/firstclassmodules.etex

```diff
@@ -1,92 +1,107 @@
+a sorted set, and then reads them back in order:
+\begin{caml_example}{verbatim}
+let sort0 (type a) li =
+  let module Set = Set.Make(struct type t = a let compare = Stdlib.compare end) in
+  Set.elements (List.fold_right Set.add li Set.empty)
 
-The @package-type@ syntactic class appearing in the  @'(' 'module'
-package-type ')'@ type expression and in the annotated forms represents a
-subset of module types.
-This subset consists of named module types with optional constraints
-of a limited form: only non-parametrized types can be specified.
+let example = sort0 [3; 1; 2]
+\end{caml_example*}
 
-For type-checking purposes (and starting from OCaml 4.02), package types
-are compared using the structural comparison of module types.
+Using module-dependent functions, it is also possible to express
+"sort" as a function parametrized over an ordered-set module; this
+lets caller choose their ordered-set implementation, which may be
+specialized for their type and for example use another comparison
+function.
 
-In general, the module expression @'(' "val" expr ":" package-type
-')'@ cannot be used in the body of a functor, because this could cause
-unsoundness in conjunction with applicative functors.
-Since OCaml 4.02, this is relaxed in two ways:
-if @package-type@ does not contain nominal type declarations ({\em
-  i.e.} types that are created with a proper identity), then this
-expression can be used anywhere, and even if it contains such types
-it can be used inside the body of a generative
-functor, described in section~\ref{s:generative-functors}.
-It can also be used anywhere in the context of a local module binding
-@'let' 'module' module-name '=' '(' "val" expr_1 ":" package-type ')'
- "in" expr_2@.
+\begin{caml_example}{verbatim}
+let sort (module Set : Set.S) li =
+  Set.elements (List.fold_right Set.add li Set.empty)
+\end{caml_example}
+
+The declaration of "sort" uses the pattern syntax "(module M : S)",
+which allows function parameters to be modules instead of ordinary
+values. We call "sort" a \emph{module-dependent} function because in
+its type, "(module Set : Set.S) -> Set.elt list -> Set.elt list", the
+module parameter appears in the right-hand-side (return type) of the
+arrow "(module Set : Set.S) -> ...", via occurrences of the type
+"Set.elt".
+
+To apply the "sort" function, we use the construction "(module M : S)"
+to provide the argument, which can typically be written "(module M)"
+as the signature can be inferred from the function.
+
+\begin{caml_example*}
+module ISetIncr = Set.Make(Int)
+module ISetDecr =
+  Set.Make(struct
+    type t = int
+    let compare i1 i2 = Int.compare i2 i1
+  end)
+\end{caml_example*}
+
+\begin{caml_example}[toplevel]
+let example = sort (module ISetIncr) [3; 1; 2]
+let example = sort (module ISetDecr) [3; 1; 2]
+
+(* Note: any module path is accepted, including functor applications *)
+let example = sort (module Set.Make(Int)) [3; 1; 2]
```

Comment:
```ocaml
let example = sort (module Set.Make(struct type t = int let compare = Int.compare end))
```
is also accepted while
```ocaml
module F () : Set.S = Set.Make(Int)
let example = sort (module F())
```
is rejected

Reply:
I don't know why `Make(struct .. end)` is accepted (it looks a bit fishy to me!), but I just removed the comment to avoid the difficulty. Thanks!